Add blog post: Immutability patterns using CEL (#36032)
kubernetes/website: content/en/blog/_posts/2022-08-15-immutability-with-cel.md

---
layout: blog
title: "Enforce CRD Immutability with CEL Transition Rules"
date: 2022-08-15
slug: enforce-immutability-using-cel
canonicalUrl: https://www.kubernetes.dev/blog/2022/08/15/enforce-immutability-using-cel
---

> **Author:** [Alexander Zielenski (Google)](https://github.com/alexzielenski)

Reviewer: ```suggestion
**Author:** [Alexander Zielenski](https://github.com/alexzielenski) (Google)
```

---

> # Introduction

Reviewer: ```suggestion
```

We don't start blog articles with a heading.

> Immutable fields can be found in a few places in the built-in Kubernetes types such as PodSpec, EphemeralContainers, Finalizers, and more.

Reviewer: ```suggestion
Immutable fields can be found in a few places in the built-in Kubernetes types.
For example, you can't change the `.metadata.name` of an object. Specific objects
have fields where changes to existing objects are constrained; for example, the
`.spec.selector` of a Deployment.
```

You _can_ mutate finalizers and ephemeral containers for an existing Pod, and to suggest otherwise is liable to confuse folk.

> Until recently the best way to create immutable fields for Custom Resource Definitions has been to create an admission webhook: an unfortunate tradeoff for the common case of immutable fields.

Reviewer: ```
Until recently the best way to restrict field mutability  for CustomResourceDefinitions
has been to create a validating
[admission webhook](/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks):
this means a lot of complexity for the common case of making a field immutable.
```

Beta since Kubernetes 1.25, CEL Validation Rules allow CRD authors to express validation constraints on their fields using a rich expression language, [CEL](https://github.com/google/cel-spec). This article explores how you can use validation rules to implement a few common immutability patterns.

> # Validation Rules Basics

Reviewer: ```suggestion
## Basics of validation rules
```

The new support for CEL validation rules in Kubernetes allows CRD authors to add complicated admission logic for their resources without writing any code!

For example, A CEL rule to constrain a field `maximumSize` to be greater than a `minimumSize` for a CRD might look like the following:
```yaml
   rule: self.maximumSize > self.minimumSize
message: “Maximum size must be greater than minimum size.”
```
The rule field contains an expression written in CEL. `self` is a special keyword in CEL which refers to the object whose type contains the rule.

The message field is an error message which will be sent to Kubernetes clients whenever this particular rule is not satisfied.

> ## More Information
> For more details about the capabilities and limitations of Validation Rules using CEL, please refer to [the documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules). The [cel-spec](https://github.com/google/cel-spec) is also a good reference for information specifically related to the language.
> 
> # Immutability Patterns with CEL Validation Rules
> This section implements several common use cases for immutability in Kubernetes Custom Resource Definitions using validation rules expressed as [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/crd.html).
> 
> We will also show the resultant OpenAPI generated by the kubebuilder marker comments so that those not using kubebuilder can also follow along.

Reviewer: ```suggestion
For more details about the capabilities and limitations of Validation Rules using CEL, please refer to [validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules). The [CEL specification](https://github.com/google/cel-spec) is also a good reference for information specifically related to the language.

## Immutability patterns with CEL validation rules

This section implements several common use cases for immutability in Kubernetes
CustomResourceDefinitions, using validation rules expressed as [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/crd.html).

I will also show the resultant OpenAPI generated by the kubebuilder marker comments so that those not using kubebuilder can also follow along.
```

## Project Setup

To demonstrate how to use CEL rules with kubebuilder comments, we first need to set up a Golang project structure with the CRD defined in Go. Afterwards, we will be able to use this go project to demonstrate different design patterns.

You may skip this step if you are only interested in the resultant OpenAPI extensions.

We begin with a folder structure of a Go module set up like the following. If you have your own project already set up feel free to adapt this tutorial to your liking:
```console
cel-immutability-tutorial
├── generate.go
├── pkg
│  └── apis
│     └── stable.example.com
│        └── v1
│           ├── doc.go
│           └── types.go
└── tools.go

> ```

Author: I looked into this a bit and I'm not sure how to make the structure more readable using mermaid. AS far as I can tell mermaid has no functionality for a folder diagram (https://github.com/mermaid-js/mermaid/issues/2645).

Attempts to approximate `tree-style` output with flow-style diagrams I think complicate the picture: https://gitlab.com/joaommpalmeiro/diagram-scratchpad/-/blob/master/README.md.

@sftim Is this what you had in mind or something else?

> ```

Reviewer: OK, fair enough. Could you make a simple SVG instead, then? We can't rely on line-drawing characters rendering how you'd like them to.

If you have a terminal emulator app that lets you save a PDF output and then convert it, that's one option.

> ```

Author: After trying a few things out I was not happy with, I capitulated and added mermaid diagrams. Sorry for the shuffle.

├── pkg
│  └── apis
│     └── stable.example.com
│        └── v1
│           ├── doc.go
│           └── types.go
└── tools.go
```

Our test CRD yaml is now available in `crds/stable.example.com_testcrds.yaml`

## Immutable After First Write
A common immutability design pattern is to make the field immutable once it has been first set. This example will throw a validation error if the field after changes after being first initialized.

```go
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.value) || has(self.value)", message="Value is required once set"
type ImmutableSinceFirstWrite struct {
   metav1.TypeMeta   `json:",inline"`
   metav1.ObjectMeta `json:"metadata,omitempty"`

   // +kubebuilder:validation:Optional
   // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
   // +kubebuilder:validation:MaxLength=512
   Value string `json:"value"`
}
```

The `+kubebuilder` directives in the comments inform controller-gen how to annotate the generated OpenAPI. The `XValidation` rule causes the rule to appear among the  `x-kubernetes-validations` OpenAPI extension. Kubernetes then respects the OpenAPI spec to enforce our constraints.

> To be immutable after first write, we require the following constraints:

Reviewer: ```suggestion
To be immutable after first write, we need to apply the following constraints:

```

$ kubectl apply -f - <<EOF
---
apiVersion: stable.example.com/v1
kind: MapAppendOnlyKeys
metadata:
  name: testmap
values:
    key1: value1
EOF

The MapAppendOnlyKeys "testmap" is invalid: values: Invalid value: "object": Keys may not be removed and their values must stay the same

$ kubectl apply -f - <<EOF
---
apiVersion: stable.example.com/v1
kind: MapAppendOnlyKeys
metadata:
  name: testmap
EOF

The MapAppendOnlyKeys "testmap" is invalid: <nil>: Invalid value: "object": Value is required once set
```

### Generated openAPIV3 Schema

```yaml
openAPIV3Schema:
  description: A map which does not allow keys to be removed or their values
    changed once set. New keys may be added, however.
  properties:
    values:
      additionalProperties:
        type: string
      maxProperties: 10
      type: object
      x-kubernetes-validations:
      - message: Keys may not be removed and their values must stay the same
        rule: oldSelf.all(key, key in self && self[key] == oldSelf[key])
  type: object
  x-kubernetes-validations:
  - message: Value is required once set
    rule: '!has(oldSelf.values) || has(self.values)'
```

# Going Further

> The above examples showed how CEL rules can be added to kubebuilder types. The same rules can be added directly to OpenAPI if writing CRD yaml by hand.

Reviewer: ```suggestion
The above examples showed how CEL rules can be added to kubebuilder types. The same rules can be added directly to OpenAPI if writing a manifest for a CRD by hand.
```

For native types, the same behavior can be achieved using kube-openapi’s marker [`+validations`](https://github.com/kubernetes/kube-openapi/blob/923526ac052c59656d41710b45bbcb03748aa9d6/pkg/generators/extension.go#L69).

> We have only scratched the surface of the expressiveness of validation rules using CEL. For more information please check out [the documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules).

Reviewer: ```suggestion
We have only scratched the surface of the expressiveness of validation rules using CEL. For more information please check out [validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) in the Kubernetes documentation.
```

    changed once set. New keys may be added, however.
  properties:
    values:
      additionalProperties:
        type: string
      maxProperties: 10
      type: object
      x-kubernetes-validations:
      - message: Keys may not be removed and their values must stay the same
        rule: oldSelf.all(key, key in self && self[key] == oldSelf[key])
  type: object
  x-kubernetes-validations:
  - message: Value is required once set
    rule: '!has(oldSelf.values) || has(self.values)'
```

# Going further
The above examples showed how CEL rules can be added to kubebuilder types.
The same rules can be added directly to OpenAPI if writing a manifest for a CRD by hand.

For native types, the same behavior can be achieved using kube-openapi’s marker
[`+validations`](https://github.com/kubernetes/kube-openapi/blob/923526ac052c59656d41710b45bbcb03748aa9d6/pkg/generators/extension.go#L69).

Usage of CEL within Kubernetes Validation Rules is so much more powerful than
what has been shown in this article. For more information please check out
[validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)

> in the Kubernetes documentation.

Reviewer: We could also link to the blog that https://github.com/kubernetes/website/pull/35459 will publish, _and_ we could amend that article to link to this one (would need us to delay merging this PR, so that we don't link to a not-yet-published article).

Author: Link to a prospective URL added. Agreed hold until link is confirmed

Author: added a link to https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/
